### PR TITLE
Improve Toitdoc comments

### PR DESCRIPTION
**Pull request by erikcorry**

### CODE REVIEW

File: src/common.toit [hunk 1]
```diff
@@ -18,57 +17,102 @@ You can add and remove element objects to a Window.  They will be drawn
...
+  Next time the draw method is called on the display, the area
+    formerly occupied by the element will be redrawn.
   */
-  write canvas/Canvas -> none:
-    write_ canvas
+  remove element/Element -> none
 
-  abstract write_ canvas
+  /**
+  Remove all elements from a Window.
```
Comment (floitsch) [inline src/common.toit:36]:
```suggestion
  Remove all elements from a Window.
  See $remove.
```

File: src/common.toit [hunk 2]
```diff
@@ -18,57 +17,102 @@ You can add and remove element objects to a Window.  They will be drawn
...
+  */
+  remove_all -> none
 
-  abstract invalidate -> none
+  /**
+  Called by elements that have been added to this, when they change
+    or move.
+  This will cause the area described to be redrawn, next time the draw method
+    is called on the display.
+  The coordinates are in the receiver's coordinate space.
```
Comment (floitsch) [inline src/common.toit:45]: I guess "receiver" is "this instance"?

File: src/common.toit [hunk 3]
```diff
@@ -18,57 +17,102 @@ You can add and remove element objects to a Window.  They will be drawn
...
 
+/**
+When the display is being redrawn, the area that needs updating will
+  be divided up into reasonably-sized rectangles.  Each rectangle
+  is represented by an object of this class.
+The draw methods of the elements will be called with this canvas, and
+  they can use methods on the canvas to draw themselves.
+The canvas will generally be smaller than the display in order to
+  reduce peak memory usage.  This means the draw method can be called
+  many times on each element.
```
Comment (floitsch) [inline src/common.toit:57]:
```suggestion
The draw methods of the elements is called with this canvas, and
  they can use methods on the canvas to draw themselves.
The canvas is generally smaller than the display in order to
  reduce peak memory usage.  This means the draw method can be called
  many times on each element.
```

File: src/common.toit [hunk 4]
```diff
@@ -81,20 +125,80 @@ abstract class Canvas:
...
+  */
   abstract bitmap x/int y/int -> none
       --pixels/ByteArray
       --alpha/ByteArray          // 2-element byte array.
       --palette/ByteArray        // 6-element byte array.
       --source_width/int         // In pixels.
       --source_line_stride/int   // In bytes.
 
+  /**
+  See $bitmap.
```
Comment (floitsch) [inline src/common.toit:197]: Can these two functions be merged into one, with an optional `orientation`?

```suggestion
  See $(bitmap bitmap x y --pixels --alpha --palette --source_width --source_line).
```

File: src/common.toit [hunk 5]
```diff
@@ -103,6 +207,33 @@ abstract class Canvas:
...
 
+  /**
+  Draws the given 8-bit pixmap on the canvas.
+  The source pixmap has one byte per pixel, which is an index into the
+    $palette and $alpha arguments.  The order of the pixmap is as in
+    PNG, so the lines are ordered from top to bottom, and within each
+    line the bytes are ordered from left to right.
+  The $alpha argument controls which pixel indexes are transparent.  A
+    byte value of 0 means pixels with that index are transparent, and a byte
+    value of 0xff means the pixels with that index are opaque.
```
Comment (floitsch) [inline src/common.toit:218]: Does the alpha bytearray need to be 256 bytes long? Do all colors need to have an entry?

File: src/common.toit [hunk 6]
```diff
@@ -103,6 +207,33 @@ abstract class Canvas:
...
+  If the $alpha argument is shorter than the highest index in the pixmap, then
+    pixels with high indeces are opaque.
```
Comment (floitsch) [inline src/common.toit:220]: either indices, or indexes.
we are moving towards "indexes".
```suggestion
    pixels with high indexes are opaque.
```

File: src/common.toit [hunk 7]
```diff
@@ -103,6 +207,33 @@ abstract class Canvas:
...
+  The palette argument has 3 bytes per color, in rgbrgb order.  For example
```
Comment (floitsch) [inline src/common.toit:221]: Makes a bit less sense with so many values, but wrapping the palette into a class still feels nice.

File: src/common.toit [hunk 8]
```diff
@@ -103,6 +207,33 @@ abstract class Canvas:
...
+    if the pixmap uses 0 to represent transparency, 1 to represent red, and
+    2 to represent white, then the $palette would be
+    #[0, 0, 0, 0xff, 0, 0, 0xff, 0xff, 0xff] and the $alpha argument would
+    be #[0] to make the 0's transparent.
```
Comment (floitsch) [inline src/common.toit:225]:
```suggestion
    2 to represent white, then the $palette should be
    #[0, 0, 0, 0xff, 0, 0, 0xff, 0xff, 0xff] and the $alpha argument should
    be #[0] to make the 0's transparent.
```

File: src/element.toit [hunk 9]
```diff
@@ -13,7 +13,14 @@ import .common
 import font show Font
 import math
 
-abstract class Element extends ElementOrTexture_ implements Window:
+/**
+Something that can be placed on a display.  They can contain other
```
Comment (floitsch) [inline src/element.toit:17]:
```suggestion
An element that can be placed on a display.  They can contain other
```

File: src/element.toit [hunk 10]
```diff
@@ -13,7 +13,14 @@ import .common
...
+  elements, and draw themselves on Canvases.
+They can be stacked up and will be drawn from back to front, with transparency.
```
Comment (floitsch) [inline src/element.toit:19]:
```suggestion
Elements can be stacked up and are drawn from back to front, with transparency.
```

**Review by floitsch — APPROVED** [review]
LGTM.

File: src/common.toit [hunk 11]
```diff
@@ -18,57 +17,102 @@ You can add and remove element objects to a Window.  They will be drawn
...
+    canvas) will generally return a gray-scale canvas where white (0xff)
+    represents an opaque pixel, and black (0) represents a transparent pixel.
+  A canvas type that does not alow mixing and averaging of colors (eg. a
+    black/white/red canvas) will return a one-bit canvas where 1
+    represents an opaque pixel, and 0 represents a transparent pixel.
+  */
   abstract make_alpha_map -> Canvas
+  /**
+  As for $make_alpha_map, but with a padding of $padding pixels around
+    all four the edges of the canvas.
```
[the remark this answers is not in the document]
Comment (erikcorry) [reply src/common.toit]: I gave make_alpha_map a default argument instead.

File: src/common.toit [hunk 12]
```diff
@@ -18,57 +17,102 @@ You can add and remove element objects to a Window.  They will be drawn
...
+  */
   abstract make_alpha_map --padding/int -> Canvas
 
-  static DISJOINT           ::= 0  // The area and the canvas are disjoint.
-  static AREA_IN_CANVAS     ::= 1  // The area is a subset of the canvas.
-  static CANVAS_IN_AREA     ::= 2  // The canvas is a subset of the area.
-  static COINCIDENT         ::= 3  // The area and the canvas are identical.
-  static OVERLAP            ::= 4  // The areas overlap, but neither is a subset of the other.
+  /// Result from $bounds_analysis: The area and the canvas are disjoint.
+  static DISJOINT           ::= 0
```
[the remark this answers is not in the document]
Comment (erikcorry) [reply src/common.toit:96]: I prefer it this way.

File: src/common.toit [hunk 13]
```diff
@@ -81,20 +125,80 @@ abstract class Canvas:
...
   abstract rectangle x/int y/int --w/int --h/int --color/int -> none
 
+  /**
+  Draws the given text on the canvas in the given color.
+  The background of the text is not drawn, ie it is transparent.
+  The text is automatically clipped to the area of the canvas
+    so it is not an error for the text to be outside the
+    canvas.
+  The orientation is normally ORIENTATION_0 (from "import bitmap"), but can be
+    ORIENTATION_90, ORIENTATION_180, or ORIENTATION_270, representing
```
[the remark this answers is not in the document]
Comment (erikcorry) [reply src/common.toit]: Does that work when they are from the SDK, not the package?  I am trying your suggestion, let's see how it works.

File: src/common.toit [hunk 14]
```diff
@@ -81,20 +125,80 @@ abstract class Canvas:
...
   abstract text x/int y/int --text/string --color/int --font/Font
 
+  /**
+  Draws the given bitmap on the canvas.
+  The bit order is as in PNG, so the lines are ordered from top to
+    bottom, and within each line the bytes are ordered from left
+    to right.  Within each byte, the high bits are on the left, and
+    the low bits are on the right.
+  Using the $alpha argument, the bitmap can be drawn with transparency.
+    For examples if $alpha is #[0, 0xff] then the zeros in the bitmap
```
[the remark this answers is not in the document]
Comment (erikcorry) [reply src/common.toit:181]: Updated to clarify that opacities that are not 0 or 0xff are allowed.

I don't want to have to create an instance to call this method, so I'm not in favour of the BitmapAlpha class.

This is not in any case a method that most users of the library will need to implement or call.  You currently use it by making a 1-bit PNG, but at some point there will likely be implementations of Background that let you specify a (tiled) bitmap, and we can look at the documentation of that.

File: src/common.toit (hunk 5 above)
Comment (erikcorry) [reply src/common.toit:215]: This is answered in the line below.

File: src/common.toit (hunk 6 above)
Comment (erikcorry) [reply src/common.toit]: I like indices, I just can't spell.

**Review by erikcorry — COMMENTED** [review]
Comments I didn't respond to, I implemented.  I'm not in favour of adding new classes for the alpha and palette.

**Pull request merged**